File: bees.lua
local createPoorQualityQueen = require("lib.advanced_mutatron.createQueen")
local finishQueen = require("lib.advanced_mutatron.finishQueen")
local killQueen = require("lib.industrial_apiary.killQueen")
local maintainMutatron = require("lib.advanced_mutatron.maintenance")
local maintainMutagen = require("lib.mutagen_producer.maintenance")
local maintainApiary = require("lib.industrial_apiary.maintenance")
local breeding = require("lib.utils.breeding")

local unknownBees = breeding.findUnknownBees()
if #unknownBees > 0 then
    print("Unknown bees:")
    for _, bee in ipairs(unknownBees) do
        print(bee)
    end
end


-- DEBUGGING
-- local allBees = beeUtils.buildBeeInventory()
-- print("Current species in all storage:")
-- for species, _ in pairs(allBees.existingSpecies) do
--     print("  " .. species)
-- end

-- DEBUGGING
-- local permBees = beeUtils.buildBeeInventory(false)
-- print("Current species in permanent storage:")
-- for species, _ in pairs(permBees.existingSpecies) do
--     print("  " .. species)
-- end

while true do
    maintainMutatron()
    maintainMutagen()
    maintainApiary()

    createPoorQualityQueen()
    finishQueen() -- TODO: Add a config setting to turn this off for players who don't want to dupe bees
    killQueen()

    sleep(1)
end

File: src/processes/bees.ts
import { wantedBees } from "../config/meatballcraft";
import { getExistingBees } from "../lib/storage-bees";
import { getBreedingPaths } from "../utils/breeding";
import { yieldSmart } from "../utils/yield";

const createQueen = () => {
    const mutatrons = peripheral.find<InventoryPeripheral>("gendustry:mutatron_advanced");
    if (mutatrons.length === 0) {
        print("No advanced mutatron found");
        return false;
    }

    const mutatron = mutatrons[0];
    const slot1 = mutatron.getItemMeta(1);
    const slot2 = mutatron.getItemMeta(2);

    if (slot1 || slot2) {
        return false; // wait for mutatron to be empty
    }

    const inventory = getExistingBees(true, true);

    for (const targetSpecies of wantedBees) {
        if (!inventory.existingSpecies[targetSpecies]) {
            print("Need: " + targetSpecies);
            const breedingPaths = getBreedingPaths();
            const allSpecies = Object.keys(breedingPaths);
            for (const species of allSpecies) {
                const allMethodsToCreateSpecies = breedingPaths[species];
                yieldSmart();
                for (const method of allMethodsToCreateSpecies) {
                    const [parent1, parent2] = method;
                    if (inventory.princesses[parent1] && inventory.drones[parent2]) {
                        print("  Creating " + species);
                        print("    Princess: " + parent1);
                        print("    Drone: " + parent2);

                        inventory.princesses[parent1].storage.pushItems(peripheral.getName(mutatron), inventory.princesses[parent1].slot, 1, 1);
                        inventory.drones[parent2].storage.pushItems(peripheral.getName(mutatron), inventory.drones[parent2].slot, 1, 2);
                        return true;
                    } else if (inventory.drones[parent1] && inventory.princesses[parent2]) {
                        print("  Creating " + species);
                        print("    Princess: " + parent2);
                        print("    Drone: " + parent1);

                        inventory.princesses[parent2].storage.pushItems(peripheral.getName(mutatron), inventory.princesses[parent2].slot, 1, 1);
                        inventory.drones[parent1].storage.pushItems(peripheral.getName(mutatron), inventory.drones[parent1].slot, 1, 2);
                        return true;
                    }
                }
            }

        }
    }
    return false;
};

const finishQueen = () => {

}

const killQueen = () => {

}

export const createNewSpecies = () => {
    createQueen()
    finishQueen() // TODO: Add a config setting to turn this off for players who don't want to dupe bees
    killQueen()
}